archive/walk.go

```diff
@@ -273,110 +224,91 @@ func (l LogID) Path(ark *Archive) iosrc.URI {
-// and last record timestamps.
-func (si SpanInfo) Span() nano.Span {
-	return nano.Span{Ts: si.First, Dur: 1}.Union(nano.Span{Ts: si.Last, Dur: 1})
+func (c Chunk) Span() nano.Span {
+	return closedSpan(c.First, c.Last)
 }
 
-func (si SpanInfo) Range(ark *Archive) string {
-	return fmt.Sprintf("[%d-%d]", si.First, si.Last)
+func (c Chunk) LogID() LogID {
+	return LogID(path.Join(dataDirname, newTsDir(c.First).name(), fmt.Sprintf("%s-%s.zng", c.DataFileKind, c.Id)))
 }
 
-type SpanVisitor func(si SpanInfo, zardir iosrc.URI) error
-
-func SpanWalk(ctx context.Context, ark *Archive, v SpanVisitor) error {
-	return tsDirVisit(ctx, ark, nano.MaxSpan, func(_ tsDir, spans []SpanInfo) error {
-		for _, s := range spans {
-			zardir := LogToZarDir(s.LogID.Path(ark))
-			if dirmkr, ok := ark.dataSrc.(iosrc.DirMaker); ok {
-				if err := dirmkr.MkdirAll(zardir, 0700); err != nil {
-					return err
-				}
-			}
-			if err := v(s, zardir); err != nil {
-				return err
-			}
-		}
-		return nil
-	})
+// ZarDir returns a URI for a directory specific to this data file, expected
+// to hold microindexes or other files associated with this chunk's data.
+func (c Chunk) ZarDir(ark *Archive) iosrc.URI {
+	return ark.DataPath.AppendPath(string(c.LogID() + zarExt))
 }
 
-// RmDirs descends a directory hierarchy looking for zar dirs and remove
-// each such directory and all of its contents.
-func RmDirs(ctx context.Context, ark *Archive) error {
-	fn := func(u iosrc.URI) error {
-		return ark.dataSrc.RemoveAll(ctx, u)
+// Localize returns a URI that joins the provided relative path name to the
+// zardir for this chunk. The special name "_" is mapped to the path of the
+// data file for this chunk.
+func (c Chunk) Localize(ark *Archive, pathname string) iosrc.URI {
+	if pathname == "_" {
+		return c.Path(ark)
 	}
-	return Walk(ctx, ark, fn)
+	return c.ZarDir(ark).AppendPath(pathname)
 }
 
-type multiSource struct {
-	ark   *Archive
-	paths []string
+func (c Chunk) Path(ark *Archive) iosrc.URI {
+	return ark.DataPath.AppendPath(string(c.LogID()))
 }
 
-// NewMultiSource returns a driver.MultiSource for an Archive. If no paths are
-// specified, the MultiSource will send a source for each chunk file, and
-// report the same ordering as the archive. Otherwise, the sources come from
-// localizing the given paths to each chunk's directory (recognizing "_" as the
-// chunk file itself), with no defined ordering.
-func NewMultiSource(ark *Archive, paths []string) driver.MultiSource {
-	if len(paths) == 0 {
-		paths = []string{"_"}
-	}
-	return &multiSource{
-		ark:   ark,
-		paths: paths,
-	}
+func (c Chunk) Range(ark *Archive) string {
+	return fmt.Sprintf("[%d-%d]", c.First, c.Last)
 }
 
-func (ams *multiSource) OrderInfo() (string, bool) {
-	if len(ams.paths) == 1 && ams.paths[0] == "_" {
-		return "ts", ams.ark.DataSortDirection == zbuf.DirTimeReverse
+func chunkTsCompare(dir zbuf.Direction, iTs nano.Ts, iKid ksuid.KSUID, jTs nano.Ts, jKid ksuid.KSUID) bool {
```

Comment:
Nit: Compare functions usually return an int (like bytes.Compare), but this is a less function (like sort.Interface.Less), so I think chunkTsLess would be a better name.